app/controllers/pieces_controller.rb

```diff
@@ -18,20 +18,31 @@ def piece_params
   def checks_before_move
     destination_x = piece_params[:column_coordinate].to_i
     destination_y = piece_params[:row_coordinate].to_i
-    if !@piece.obstructed?(destination_x, destination_y) && @piece.valid_move?(destination_x, destination_y)
+    render plain: 'Invalid Move', status: :bad_request if not_moved?(destination_x, destination_y)
+    if !@piece.obstructed?(destination_x, destination_y) && @piece.valid_move?(destination_x, destination_y) && check_turn == @piece.color
+      change_turn
       @piece.move_to!(destination_x, destination_y)
-      checks_turn_and_color
+      @piece.game.update(turn: -1) if @piece.game.checkmate?
+      flash[:notice] = 'Checkmate !!' if @piece.game.turn == -1
       redirect_to game_path(@piece.game_id)
     else
       render plain: 'Invalid Move', status: :bad_request
     end
   end
 
-  def checks_turn_and_color
-    if current_game.turn == current_game.white_player.id && @piece.color == 'white'
-      current_game.update_attributes(turn: current_game.black_player_id)
-    elsif current_game.turn == current_game.black_player.id && @piece.color == 'black'
-      current_game.update_attributes(turn: current_game.white_player_id)
-    end
+  def not_moved?(destination_x, destination_y)
+    return true if destination_x == @piece.column_coordinate && destination_y == @piece.row_coordinate
+    false
+  end
```

Comment:
Generally it's best to keep "logic" like this in the model. See if you can move it into a model and use it from there.